client/core/core.go

```diff
@@ -1052,11 +1052,28 @@ func (c *Core) connectedWallet(assetID uint32) (*xcWallet, error) {
 	return wallet, nil
 }
 
+// connectWallet connects to wallet and validate the known deposit address
+// after successfull connectection, if deposit address does not belong to
+// wallet it generates new address and updates xcWallet and dbWallet, therefore
+// it might hold the wallet lock
 func (c *Core) connectWallet(w *xcWallet) error {
 	err := w.Connect(c.ctx)
 	if err != nil {
 		return codedError(connectWalletErr, err)
 	}
+	// Check if known address belongs to connected wallet
+	mine, err := w.ValidateAddress(w.address)
+	if err != nil {
+		return err
+	}
+	if !mine {
+		c.log.Infof("[%v]: Deposit address %v does not belong to connected wallet"+
+			", generating new address", unbip(w.AssetID), w.address)
+		_, err := c.NewDepositAddress(w.AssetID)
```

Comment:
I'm also noting what appears to be an awkward use pattern that is making issues (or at least confusion with the walletMtx). `(*Core).NewDepositAddress(uint32)` was previously only consumed by the webserver api for generating new addresses, so it included a lookup of an `xcWallet` by id in the `wallets` map, thus necessitating locking `walletsMtx` for the map access.  However, here in `connectWallet` you already have the target `wcWallet`.  I would factor out the guts of `NewDepositAddress` into an unexported `newDepositAddress(w *xcWallet`)`, which you would use here.  The exported version would remain, doing a lookup by id and then calling the unexported one.

Reply:
Sounds good, will do, thanks 👍